lib/stoplight/data_store/redis.rb

```diff
@@ -94,16 +92,50 @@ def clear_state(config)
         normalize_state(state)
       end
 
-      LOCK_TTL = 2_000 # milliseconds
-
-      def with_notification_lock(config, from_color, to_color)
-        @redlock.lock(notification_lock_key(config), LOCK_TTL) do
-          if last_notification(config) != [from_color, to_color]
-            set_last_notification(config, from_color, to_color)
+      NOTIFICATION_DEDUPLICATION_TTL = 60 # TTL for notification deduplication (in seconds)
+      private_constant :NOTIFICATION_DEDUPLICATION_TTL
 
-            yield
-          end
+      # This Lua script implements a notification deduplication mechanism:
+      # 1. It checks if the current color transition (from_color → to_color) matches the previously recorded one
+      # 2. If it's the same transition, returns 0 (don't notify) to prevent duplicate notifications
+      # 3. If it's a different transition (or first time), it:
+      #    - Records the new transition with a timestamp
+      #    - Sets an expiration time on the record (allowing notifications to repeat after TTL as a fail-safe mechanism)
+      #    - Returns 1 (proceed with notification)
+      #
+      # This ensures that when multiple servers detect the same transition simultaneously,
+      # only one notification is sent, while still allowing future transitions to trigger notifications.
+      #
+      # This script is executed on redis and it's guaranteed that the operations are atomic.
+      NOTIFICATION_DEDUPLICATION_SCRIPT = <<~LUA
```

Comment:
If we use more LUA scripts in the future, I'd place them separately from the `data_store/redis.rb` file (e.g., `data_store/lua_scripts` or something).

Reply:
This is a good idea. We can use the Lua script to store errors as well. Let's keep it as is for now, but this is definitely a way to go if we have more such scripts